File: juno_rs/src/stop_loss/trailing.rs
use super::StopLoss;
use crate::{genetics::Chromosome, Candle};
use juno_derive_rs::*;
use rand::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Chromosome, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct TrailingParams {
    pub up_threshold: f64,
    pub down_threshold: f64,
}

fn up_threshold(rng: &mut StdRng) -> f64 {
    rng.gen_range(0.001..1.000)
}
fn down_threshold(rng: &mut StdRng) -> f64 {
    up_threshold(rng)
}

pub struct Trailing {
    up_threshold_factor: f64,
    down_threshold_factor: f64,
    highest_close_since_position: f64,
    lowest_close_since_position: f64,
    close: f64,
}

impl Trailing {
    pub fn new(params: &TrailingParams) -> Self {
        Self {
            up_threshold_factor: 1.0 - params.up_threshold,
            down_threshold_factor: 1.0 + params.down_threshold,
            highest_close_since_position: 0.0,
            lowest_close_since_position: f64::MAX,
            close: 0.0,
        }
    }
}

impl StopLoss for Trailing {
    fn upside_hit(&self) -> bool {
        self.close <= self.highest_close_since_position * self.up_threshold_factor
    }

    fn downside_hit(&self) -> bool {
        self.close >= self.lowest_close_since_position * self.down_threshold_factor
    }

    fn clear(&mut self, candle: &Candle) {
        self.highest_close_since_position = candle.close;
        self.lowest_close_since_position = candle.close;
    }

    fn update(&mut self, candle: &Candle) {
        self.close = candle.close;
        self.highest_close_since_position =
            f64::max(self.highest_close_since_position, candle.close);
        self.lowest_close_since_position = f64::min(self.lowest_close_since_position, candle.close);
    }
}

File: juno/stop_loss/trailing.py
from decimal import Decimal
from typing import Optional

from juno import Candle

from .stop_loss import StopLoss


class Trailing(StopLoss):
    _up_threshold_factor: Decimal
    _down_threshold_factor: Decimal
    _highest_close_since_position = Decimal("0.0")
    _lowest_close_since_position = Decimal("Inf")
    _close: Decimal = Decimal("0.0")

    def __init__(self, up_threshold: Decimal, down_threshold: Optional[Decimal] = None) -> None:
        if down_threshold is None:
            down_threshold = up_threshold
        assert 0 <= up_threshold <= 1
        assert 0 <= down_threshold <= 1
        self._up_threshold_factor = 1 - up_threshold
        self._down_threshold_factor = 1 + down_threshold

    @property
    def upside_hit(self) -> bool:
        return self._close <= self._highest_close_since_position * self._up_threshold_factor

    @property
    def downside_hit(self) -> bool:
        return self._close >= self._lowest_close_since_position * self._down_threshold_factor

    def clear(self, candle: Candle) -> None:
        self._highest_close_since_position = candle.close
        self._lowest_close_since_position = candle.close

    def update(self, candle: Candle) -> None:
        self._close = candle.close
        self._highest_close_since_position = max(self._highest_close_since_position, candle.close)
        self._lowest_close_since_position = min(self._lowest_close_since_position, candle.close)
